### PR TITLE
Update SageMaker Training Compiler MNMG Example Notebook for PT1.11

### PR DESCRIPTION
**Pull request by BruceChang9688**
*Issue #, if available:*

*Description of changes:*
As title, update SageMaker Training Compiler Multi-Node Multi-GPUs Example notebook and test file for PT1.11

*Testing done:*
[language-modeling-1018-10.pdf](https://github.com/aws/amazon-sagemaker-examples/files/9815625/language-modeling-1018-10.pdf)

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [x] I have linted my notebook(s) and code using `black-nb -l 100 {path}/{notebook-name}.ipynb`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/scripts/run_clm_smddp.py [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+    TrainingArguments,
+    default_data_collator,
+    set_seed,
+)
+from transformers.trainer_utils import get_last_checkpoint
+from transformers.utils import check_min_version
+from transformers.utils.versions import require_version
+
+# for enable herring
+import smdistributed.dataparallel.torch.torch_smddp
```
Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/scripts/run_clm_smddp.py:53]: SMDDP is fully integrated with` transformers`. This should not be necessary.
`transformers` automatically [checks for SMDDP](https://github.com/huggingface/transformers/blob/3e2dd7f92d20f67e34f649cdafbbe28d22a75f44/src/transformers/utils/import_utils.py#L594-L605) and does the [required setup](https://github.com/huggingface/transformers/blob/b79028f0b69886e2a9238482c848aa5556d890b2/src/transformers/training_args.py#L1395-L1396).

**Review by Lokiiiiii — CHANGES_REQUESTED** [review]

File: sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb [hunk 2]
```diff
@@ -2,15 +2,15 @@
  "cells": [
```
Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #18.    INSTANCE_TYPE = "ml.p4d.24xlarge"  # ml.p3.8xlarge is easily available. However, ml.p3.16xlarge provides better performance.
</pre><p>Remove the comment about P3 instances</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>AMP =&gt; Automatic Mixed Precision</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>Ensure all the HPs set in https://github.com/aws/amazon-sagemaker-examples/blob/main/sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_single_node/language-modeling/gpt-2.ipynb are set here. I see a few missing ones - dataloader_drop_last, evaluation_strategy, etc.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #10.        "do_eval": True,
</pre><p>Disable evaluation.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #30.        transformers_version="4.21.1",
</pre><p>PyTorch estimator does not support this argument</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #4.    hyperparameters["sagemaker_pytorch_xla_multi_worker_enabled"] = True
</pre><p>Please use the distribution parameter in the estimator. Refer to https://github.com/aws/amazon-sagemaker-examples/blob/main/sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_single_node/language-modeling/gpt-2.ipynb</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #7.    hyperparameters["per_device_train_batch_size"] = 84
</pre><p>The learning rate needs to be adjusted with batch size</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #11.        entry_point=f"run_{LANGUAGE_MODELING_LOSS}.py",  # Wrapper around training script that enables multi node training
</pre><p>Remove stale comment about wrapper script.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #19.        transformers_version="4.21.1",
</pre><p>Use major versions 1.11 and 4.21</p><p>This offers more flexibility while doing minor releases.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p><a href="https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/#Training-with-Native-PyTorch" target="_blank">Training with Native PyTorch</a> -&gt; <a href="https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/#Training-with-Native-PyTorch" target="_blank">Training with Native PyTorch</a> + SM DDP</p><p> </p><p><a href="https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/#Training-with-Optimized-PyTorch" target="_blank">Training with Optimized PyTorch</a> -&gt; <a href="https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/#Training-with-Optimized-PyTorch" target="_blank">Training with </a>SageMaker Training Compiler</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>Split estimator args into a different cell.</p><p>Add some markdown before the estimator config- "First, we define some basic parameters common to all estimators.</p><p><strong>Note</strong>: We recommend you to turn the SageMaker Debugger's profiling and debugging tools off to avoid additional overheads."</p><p>Check https://github.com/aws/amazon-sagemaker-examples/blob/main/sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_single_node/language-modeling/gpt-2.ipynb</p><p> </p><p>Do the same for the common hyperparameters as well</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <h3>Training with SageMaker Distributed Data Parallel</h3><h3> -&gt; Training with Native PyTorch + SM DDP</h3><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>move this cell outside the SM DDP section, this is common to both flavors of estimators here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #17.        },  # Use SageMaker Data Distributed Parallel to train across nodes/GPUs.
</pre><p>SageMaker Data Distributed Parallel -&gt; SageMaker Distributed Data Parallel</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>The note needs to correspond with the cell where we turn of debugger &amp; profiler. This has been moved to a common block above.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #10.    plt.xticks(ticks=[1, 1.5], labels=["Baseline PT", "Compiler-enhanced PT"])
</pre><p>Baseline PT -&gt; PT + SM DDP</p><p>Compiler-enhanced PT -&gt; PT + Compiler</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <pre class="ql-syntax" spellcheck="false">Line #3.    pd.DataFrame([n["summary"], o["summary"]], index=["Native", "Compiler-enhanced"])
</pre><p>Native -&gt; PT + SM DDP</p><p>Compiler-enhanced -&gt; PT + Compiler</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>Native -&gt; PT + SM DDP</p><p>Optimized -&gt; PT + Compiler</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (Lokiiiiii) [reply sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>Fix the markdown formatting.</p>

File: sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb [hunk 3]
```diff
@@ -2,30 +2,30 @@
...
```
Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>Provide justification for why we need these libraries in this environment. I don't see <code>torch</code> being used anywhere. Same for <code>s3fs.</code> Please prune this list.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>Why do we need datasets ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>Missing note about <code>distribution={"pytorchxla": {"enabled": True}}</code>.</p><p>refer to multi GPU example.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/huggingface/pytorch_multiple_gpu_multiple_node/language-modeling-multi-gpu-multi-node.ipynb:2]: <p>We are not using AMP anymore.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3611/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

**Review by aaronmarkham — APPROVED** [review]
These updates look good, but we need to offline discuss the dataset.
